Fix undefined vars on python3 and a whole bunch of other cleanup.

abadger opened the pull request:
Please review, especially the potential logic errors as I don't know what the
correct code should do there.

References #27193

* refactored new_resource_to_string_convert:
  * No need to catch any exception and then reraise it. Just let it continue to
    bubble up until caught.  Catching it and reraising it just serves to
    obfusscate the traceback which will make any problems much harder to debug

* refactored object_to_string_new:
  * No need to catch exception and reraise.  This just obfuscates the traceback
  * Build up a list and then join at the end instead of building up a string.
    list.append() is faster than string concatenation
  * No need to extract k, v pairs from one dict to make a second dict and then
    extract k, v pairs from the second dict.  Iterate over the k, v pairs
    extracted from the first dict directly instead of building the second dict.
  * No need to check if the dict is empty before iterating on it.  Iterating on
    an empty dict will automatically go to the end of the loop
  * Use isinstance instead of type(obj) is class, handles inheritance and is better style
  * use to_native instead of v.encode().  We can use the surrogate_or_strict
    error handler to deal with more potential tracebacks.  Does the right
    conversion on both Py2 and Py3.
  * Convert bool to string before combining it with the string we're building.
    This would have raised an exception as you can't add a bool to a string
  * Don't reference unicode directly as unicode does not exist in Python3
  * Convert int into a string before combining with a string we're building.
    This would have raised an exception as you can't add a bool to a string
  * Possible logical errors that I didn't change:
    * bools are not surrounded by quotes but ints are.
    * strings are not passed through cls.encode() while they are in
      object_to_string_withoutquotes_new()
    * The string resulting from this function will have a trailing comma
      whereas object_to_string_withoutquotes_new() will not have a trailing
      comma

* refactored object_to_string_withoutquotes_new:
  * No need to catch exception and reraise.  This just obfuscates the traceback
  * Build up a list and then join at the end instead of building up a string.
    list.append() is faster than string concatenation
  * No need to extract k, v pairs from one dict to make a second dict and then
    extract k, v pairs from the second dict.  Iterate over the k, v pairs
    extracted from the first dict directly instead of building the second dict.
  * No need to check if the dict is empty before iterating on it.  Iterating on
    an empty dict will automatically go to the end of the loop
  * Use isinstance instead of type(obj) is class, handles inheritance and is better style
  * use to_native instead of v.encode().  We can use the surrogate_or_strict
    error handler to deal with more potential tracebacks.  Does the right
    conversion on both Py2 and Py3.
  * Convert bool to string before combining it with the string we're building.
    This would have raised an exception as you can't add a bool to a string
  * Don't reference unicode directly as unicode does not exist in Python3
  * Possible logical errors that I didn't change:
    * strings are passed through cls.encode() while they are not passed through
      that function in object_to_string_new()
    * The string resulting from this function will not have a trailing comma
      whereas object_to_string_new() will have a trailing comma

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/netscaler.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
devel
```

abadger committed:
Fix undefined vars on python3 and a whole bunch of other cleanup.

Please review, especially the potential logic errors as I don't know what the
correct code should do there.

References #27193

* refactored new_resource_to_string_convert:
  * No need to catch any exception and then reraise it. Just let it continue to
    bubble up until caught.  Catching it and reraising it just serves to
    obfusscate the traceback which will make any problems much harder to debug

* refactored object_to_string_new:
  * No need to catch exception and reraise.  This just obfuscates the traceback
  * Build up a list and then join at the end instead of building up a string.
    list.append() is faster than string concatenation
  * No need to extract k, v pairs from one dict to make a second dict and then
    extract k, v pairs from the second dict.  Iterate over the k, v pairs
    extracted from the first dict directly instead of building the second dict.
  * No need to check if the dict is empty before iterating on it.  Iterating on
    an empty dict will automatically go to the end of the loop
  * Use isinstance instead of type(obj) is class, handles inheritance and is better style
  * use to_native instead of v.encode().  We can use the surrogate_or_strict
    error handler to deal with more potential tracebacks.  Does the right
    conversion on both Py2 and Py3.
  * Convert bool to string before combining it with the string we're building.
    This would have raised an exception as you can't add a bool to a string
  * Don't reference unicode directly as unicode does not exist in Python3
  * Convert int into a string before combining with a string we're building.
    This would have raised an exception as you can't add a bool to a string
  * Possible logical errors that I didn't change:
    * bools are not surrounded by quotes but ints are.
    * strings are not passed through cls.encode() while they are in
      object_to_string_withoutquotes_new()
    * The string resulting from this function will have a trailing comma
      whereas object_to_string_withoutquotes_new() will not have a trailing
      comma

* refactored object_to_string_withoutquotes_new:
  * No need to catch exception and reraise.  This just obfuscates the traceback
  * Build up a list and then join at the end instead of building up a string.
    list.append() is faster than string concatenation
  * No need to extract k, v pairs from one dict to make a second dict and then
    extract k, v pairs from the second dict.  Iterate over the k, v pairs
    extracted from the first dict directly instead of building the second dict.
  * No need to check if the dict is empty before iterating on it.  Iterating on
    an empty dict will automatically go to the end of the loop
  * Use isinstance instead of type(obj) is class, handles inheritance and is better style
  * use to_native instead of v.encode().  We can use the surrogate_or_strict
    error handler to deal with more potential tracebacks.  Does the right
    conversion on both Py2 and Py3.
  * Convert bool to string before combining it with the string we're building.
    This would have raised an exception as you can't add a bool to a string
  * Don't reference unicode directly as unicode does not exist in Python3
  * Possible logical errors that I didn't change:
    * strings are passed through cls.encode() while they are not passed through
      that function in object_to_string_new()
    * The string resulting from this function will not have a trailing comma
      whereas object_to_string_new() will have a trailing comma

abadger commented:
@chiradeep @giorgos-nikolopoulos could you please review this change?  We want to be able to automatically check for undefined variables and cleaning up the current codebase is needed for that.

giorgos-nikolopoulos commented on lib/ansible/module_utils/netscaler.py:
```diff
@@ -260,65 +261,43 @@ def monkey_patch_nitro_api():
+                output.append('"%s"' % v)
+            elif isinstance(v, int):
+                output.append('"%s"' % v)
+            output.append(",")
```
Discard the last comma before concatenation.
For example:
```
output.pop()
```

giorgos-nikolopoulos commented on lib/ansible/module_utils/netscaler.py:
```diff
@@ -260,65 +261,43 @@ def monkey_patch_nitro_api():
+                output.append(cls.encode(v))
+            elif isinstance(v, int):
+                output.append(str(v))
+        return ','.join(output)
```
concatenation logic should be the same as in the object_to_string_new.
Append comma at end of loop
```
output.append(',')
```
And concatenating with the empty string
```
# pop last comma
output.pop()
return ''.join(output)
```
As it is now commas are added between keys and values.

giorgos-nikolopoulos commented on lib/ansible/module_utils/netscaler.py:
```diff
@@ -32,6 +32,7 @@
 
 from ansible.module_utils.basic import env_fallback
+from ansible.module_utils.six import binary_type, text_type
 
```
Missing import
```
from ansible.module_utils._text import to_native
```

giorgos-nikolopoulos commented:
See the inline comments for code changes.

There seems to be a lot of ansibot reported errors.

Are these real issues or just false positives?

abadger committed:
Fix imports and formatting of output

abadger committed:
Simplify the conversion to string

int and bool values are now used in string formatting which will use
str to transform them without an explicit invocation.

For the unquoted function, int and bool now format the same.

abadger committed:
Remove space between key-value separator and value

k:v  instead of k: v

abadger commented:
They *were* real issues.  Fixed in the subsequent rebased commit.  (ansibot is just reflecting information from shippable into the ticket.  So you can see whether there are currently issues by checking the status of the shippable check at the bottom of the ticket).

Fixed the problems mentioned inline.  Output formats will now look like this:
**Edit** Removed space between colon and value.
```
quoted:
"key1":False,"key2":"value","key3":"3"
unquoted:
key1:False,key2:value,key3:3
```

Output from pre-PR code should have been (minus the tracebacks from adding incompatible types):
```
quoted:
"key1":False,"key2":"value","key3":"3",
unquoted:
key1:False,key2:value,key3:3
```

abadger committed:
Fix missing parenthesis

giorgos-nikolopoulos commented:
I ran the tests that are affected by the changes successfully.

It's a ship it for me.

abadger commented:
Thanks @giorgos-nikolopoulos !  merged to devel for release in 2.4.0.
